src/ol/webgl/ShaderBuilder.js

```diff
@@ -18,237 +18,669 @@ export function formatNumber(v) {
+  return getValueType(value) === ValueTypes.NUMBER;
+}
+
+/**
+ * @param {import("../style/LiteralStyle").ExpressionValue} value Either literal or an operator.
+ * @returns {boolean} True if a string value, false otherwise
+ */
+export function isValueTypeString(value) {
+  return getValueType(value) === ValueTypes.STRING || getValueType(value) === ValueTypes.COLOR_OR_STRING;
+}
+
+/**
+ * @param {import("../style/LiteralStyle").ExpressionValue} value Either literal or an operator.
+ * @returns {boolean} True if a color value, false otherwise
+ */
+export function isValueTypeColor(value) {
+  return getValueType(value) === ValueTypes.COLOR || getValueType(value) === ValueTypes.COLOR_OR_STRING;
+}
+
+
+/**
+ * Check that the provided value or expression is valid, and that the types used are compatible.
+ *
+ * Will throw an exception if found to be invalid.
+ *
+ * @param {import("../style/LiteralStyle").ExpressionValue} value Either literal or an operator.
+ */
+export function check(value) {
+  const v = value;
+
+  // these will be used to validate types in the expressions
+  function checkNumber(value) {
+    if (!isValueTypeNumber(value)) {
+      throw new Error(`A numeric value was expected, got ${JSON.stringify(value)} instead`);
+    }
+  }
+  function checkColor(value) {
+    if (!isValueTypeColor(value)) {
+      throw new Error(`A color value was expected, got ${JSON.stringify(value)} instead`);
+    }
+  }
+  function checkString(value) {
+    if (!isValueTypeString(value)) {
+      throw new Error(`A string value was expected, got ${JSON.stringify(value)} instead`);
+    }
+  }
+
+  // first check that the value is of a recognized kind
+  if (!isValueTypeColor(value) && !isValueTypeNumber(value) && !isValueTypeString(value)) {
+    throw new Error(`No type could be inferred from the following expression: ${JSON.stringify(value)}`);
+  }
+
+  // check operator arguments
+  if (Array.isArray(value) && typeof value[0] === 'string') {
+    switch (value[0]) {
+      case 'get':
+      case 'var':
+        checkString(v[1]);
+        break;
+      case 'time':
+        break;
+      case '*':
+      case '/':
+      case '+':
+      case '-':
+      case 'mod':
+      case 'pow':
+        checkNumber(v[1]);
+        checkNumber(v[2]);
+        break;
+      case 'clamp':
+        checkNumber(v[1]);
+        checkNumber(v[2]);
+        checkNumber(v[3]);
+        break;
+      case 'stretch':
+        checkNumber(v[1]);
+        checkNumber(v[2]);
+        checkNumber(v[3]);
+        checkNumber(v[4]);
+        checkNumber(v[5]);
+        break;
+      case '>':
+      case '>=':
+      case '<':
+      case '<=':
+      case '==':
+        checkNumber(v[1]);
+        checkNumber(v[2]);
+        break;
+      case '!':
+        checkNumber(v[1]);
+        break;
+      case 'between':
+        checkNumber(v[1]);
+        checkNumber(v[2]);
+        checkNumber(v[3]);
+        break;
+
+      case 'interpolate':
+        checkNumber(v[1]);
```

Comment:
I think we should use more context when checking/coercing types.  For example, when we have `color: ['interpolate', ...args]`, then we would pass along the expectation that values are color type, and when we have `size: ['interpolate', ...args]`, then we would pass along the expectation that values are numeric (I also think that we could collapse `stretch` and `interpolate` into a single `interpolate` expression a la https://docs.mapbox.com/mapbox-gl-js/style-spec/#expressions-interpolate).

Reply:
I made the assumption that each operator would only expect one type of value, in order to simplify code. Looking at the `interpolate` operator, I realized it would not really make sense to interpolate *numbers* anyway (and strings neither), so I went for colors only.

But looking again at the `interpolate` operator of Mapbox style, I guess it would indeed be a good match! If that's alright, I'll work towards this in another PR. Thanks

Reply:
I also like the idea of having typed operators, but working with colors already suggests a variadic operator (if we accept a string or rgb(a) array).  Choosing the type of interpolation (linear, exponential) can also be useful.  For example, scaling a circle radius with zoom level:

```js
['interpolate', ['exponential', 2], ['zoom'], 0, 1, 18, 12500]
```

(This could also be `linear` with `['resolution']`, but I think people have an easier time working with zoom compared to resolution.)

Anyway, I agree it makes sense to work on this separately.

Reply:
Definitely agree. `interpolate` can become a cornerstone of style expressions if done correctly (pretty much like Mapbox did it IMO). I think we should organize the code a bit differently to allow for more logic in each separate operator, e.g. to support different input/output kinds, variable argument count, etc. This would allow us to have more powerful operators, and hopefully less of them!